consensus/src/dag/shutdown.rs

```diff
@@ -0,0 +1,123 @@
+use std::{pin::Pin, task::{Context, Poll}};
+
+use futures::{Future, FutureExt};
+use tokio::{
+    select,
+    sync::{broadcast, mpsc, oneshot},
+};
+
+/// A shutdown group allows shutting down multiple async running units and
+/// await acknowledgement with one method call.
+#[derive(Debug)]
+pub(super) struct ShutdownGroup {
+    notify_shutdown: broadcast::Sender<mpsc::Sender<()>>,
+    shutdown_complete_tx: mpsc::Sender<()>,
+    shutdown_complete_rx: mpsc::Receiver<()>,
+}
+
+impl ShutdownGroup {
+
+    /// Creates a new [ShutdownGroup] from which [ShutdownHandle]s and [Shutdown]
+    /// listeners can be derived.
+    pub fn new() -> Self {
+        let (notify_shutdown, _) = broadcast::channel(1);
+        let (shutdown_complete_tx, shutdown_complete_rx) = mpsc::channel(1);
+        Self {
+            notify_shutdown,
+            shutdown_complete_tx,
+            shutdown_complete_rx,
+        }
+    }
+
+    /// Creates and returns a new pair of [ShutdownHandle] and [Shutdown] listener.
+    pub fn new_child(&self) -> (ShutdownHandle, Shutdown) {
+        let (child_notify_tx, child_notify_rx) = oneshot::channel();
+        (
+            ShutdownHandle {
+                notify_tx: child_notify_tx,
+            },
+            Shutdown {
+                is_shutdown: false,
+                global_notify: self.notify_shutdown.subscribe(),
+                single_notify: child_notify_rx,
+            },
+        )
+    }
+
+    /// Sends the shutdown signal to all child listeners and wait for them to exit.
+    pub async fn shutdown(self) {
+        let Self {
+            notify_shutdown,
+            shutdown_complete_tx,
+            mut shutdown_complete_rx,
+        } = self;
+
+        let _ = notify_shutdown.send(shutdown_complete_tx);
+        drop(notify_shutdown);
+        let _ = shutdown_complete_rx.recv().await;
```

Comment:
I am confused here.
you broadcast the sender to all, but wait only for one to response

Reply:
I see, you wait until all drop.